### PR TITLE
Issue 196 filtering

### PR DESCRIPTION
**Pull request by dandlezzz**

### CODE REVIEW

File: crates/graphql/Cargo.toml [hunk 1]
```diff
@@ -24,6 +24,7 @@ percent-encoding = "2.1.0"
 serde = { version = "1.0.130", features = ["derive"] }
 serde_json = "1.0.70"
 thiserror = "1.0.30"
+diesel = { version = "1.4.8", features = ["postgres", "r2d2", "uuidv07", "chrono", "serde_json"] }
```
Comment (kespinola) [inline crates/graphql/Cargo.toml:27]: drop yeah?

File: crates/graphql/src/schema/query_root.rs [hunk 2]
```diff
@@ -86,84 +97,17 @@ impl QueryRoot {
...
-                    .filter(
-                        attributes::trait_type
-                            .eq(trait_type)
-                            .and(attributes::value.eq(any(values))),
-                    );
-
-                acc.filter(metadatas::address.eq(any(sub)))
-            },
-        );
+        let nfts = queries::metadatas::load_filtered(
```
Comment (kespinola) [inline crates/graphql/src/schema/query_root.rs:103]:
```suggestion
        let nfts = queries::metadatas::using_filters(
```

nb.

File: crates/core/src/db/queries/metadatas.rs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            metadatas::address,
+            metadatas::name,
+            metadatas::seller_fee_basis_points,
+            metadatas::mint_address,
+            metadatas::primary_sale_happened,
+            metadata_jsons::description,
+            metadata_jsons::image,
+        ));
+
+        // let sql = debug_query::<diesel::pg::Pg, _>(&test_query);
```
Comment (kespinola) [inline crates/core/src/db/queries/metadatas.rs:84]: Adding as a comment on the PR. can remove from code.

```rust
let sql = debug_query::<diesel::pg::Pg, _>(&test_query);
let result = sql.to_string().replace("\"", "");
debug!("{:?}", result);
```

**Review by kespinola — CHANGES_REQUESTED** [review]
fmting and drop of diesel from graphql.

File: crates/graphql/Cargo.toml (hunk 1 above)
Comment (dandlezzz) [reply crates/graphql/Cargo.toml]: ya

**Review by kespinola — APPROVED** [review]

File: crates/core/src/db/queries/metadatas.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        metadatas::address,
+        metadatas::name,
+        metadatas::seller_fee_basis_points,
+        metadatas::mint_address,
+        metadatas::primary_sale_happened,
+        metadata_jsons::description,
+        metadata_jsons::image,
+    ));
+
+    let rows: Vec<Nft> = test_query.load(conn).context("failed to load nft(s)")?;
```
Comment (kespinola) [inline crates/core/src/db/queries/metadatas.rs:84]: don't need the test_query intermediate variable.

File: crates/core/src/db/queries/metadatas.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
+use std::sync::Arc;
```
Comment (ray-kast) [inline crates/core/src/db/queries/metadatas.rs:1]:
```suggestion
```

File: crates/core/src/db/queries/metadatas.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+use crate::{
+    db::{
+        models::Nft,
+        schema,
+        tables::{attributes, metadata_creators, metadata_jsons, metadatas, token_accounts},
+        Connection, Pool,
+    },
+    error::prelude::*,
+    prelude::debug,
```
Comment (ray-kast) [inline crates/core/src/db/queries/metadatas.rs:17]:
```suggestion
    prelude::*,
```

File: crates/core/src/db/queries/metadatas.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+};
+#[derive(Debug)]
+pub struct MetadataFilterAttributes {
+    pub trait_type: String,
+    pub values: Vec<String>,
+}
```
Comment (ray-kast) [inline crates/core/src/db/queries/metadatas.rs:23]: Needs docs

File: crates/core/src/db/queries/mod.rs [hunk 8]
```diff
@@ -1,4 +1,5 @@
 //! Reusable query operations for common or complicated queries.
 
 pub mod metadata_edition;
+pub mod metadatas;
```
Comment (ray-kast) [inline crates/core/src/db/queries/mod.rs:4]: `metadatas.rs` needs a `//!` self-describing doc comment at the top of the file

File: crates/graphql/src/schema/query_root.rs [hunk 9]
```diff
@@ -1,3 +1,5 @@
+use diesel::{debug_query, sql_query};
```
Comment (ray-kast) [inline crates/graphql/src/schema/query_root.rs:1]:
```suggestion
```

**Review by ray-kast — CHANGES_REQUESTED** [review]
[This region](https://github.com/holaplex/solana-indexer/pull/197/files#diff-1d97c4aa1374d415c8721ff1ff83a6b3b8220b367ab1224ecc3ad1933dc6c666R12-R13) contains unused imports but GitHub won't let me comment on it.

**Review by ray-kast — APPROVED** [review]

**Pull request merged**